### DISCUSSION TITLE
Filter Page URLs

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#136] [Anything Else]
"I have some pages that contain UUIDs that are specific to users. I don't care which date or UUID is called and want to aggregate them.

Example:
Before: /account/550e8400-e29b-11d4-a716-446655440000
After: /account/{uuid}

A simple search and replace functionality with regex would be great!"

**UPDATE:** This is now live: [Specify a custom location to aggregate pages that contain identifiers](https://plausible.io/docs/custom-locations).

**Comment by pelme** [comment]
This would be helpful from a privacy *and* analytics perspective for us.

**Comment by marcobusemann** [comment]
Same here. Would be nice to specify multiple patterns to aggregate page URLs. Currently having dynamic URL parts creates a mess in the statistic.

**Comment by aradabir007** [comment]
It is indeed creating a mess and completely unnecessary to show stats for those URLs. Would love to see this feature added anytime soon!

**Comment by jaimeiniesta** [comment]
This would be very useful to filter stats on different sections. For example, I'd like to see stats of the most visited pages of our Blog, in this case I would filter web pages starting with `http://example.com/blog`

**Reply by metmarkosaric** [reply]
Definitely! This is the use case I'd like to see too!

**Reply by captainrobbo** [reply]
I would really like this feature.  Our site has a unique "root URL" for each sporting competitions, and a forest of sub-URLs under that.    Real example:  Estonian domestic athletics championships last weekend, visible here:
     https://estonia.opentrack.run/en-gb/x/2021/EST/ec21_0223/

Ideally, one could make regex queries with filters=event:page.  So I could query for this: (excuse the non-regex)
     /[language_prefix]/x/2021/EST/ec21_0223/*
And I would get visitors to any page under or related to that championships. I could then pull this back a week later and store it.

I imagine any complex public-facing site, like a news site, would really like this feature

**Reply by metmarkosaric** [reply]
this little thread within the thread is a bit different from the main topic but probably a more widely needed addition. basically allowing you to create custom filters. right now you can filter anything by things you click on but would be great to also be able to click on say "add custom filter" which then gives you flexibility to select metric such as pages or referral sources and then select containing / not containing / exact and then you can freely type whatever you want to filter by. then you go "pages" + "containing" + "blog" to see all traffic to all pages that contain "blog" in the path.

**Reply by ukutaht** [reply]
Another thing to add to the discussion: In some cases you want to do live, ad-hoc filtering for the data already collected in your account. So nothing would change in the data we store, we would just provide more powerful tooling to apply filters on top of the existing URLs.

But if you look at the original request and the motivation some people have to get personal data out of their URLs, we probably also need to support an option where the filtering/combining is applied when the data is ingested. So instead of storing a URL `/account/123` we would in our database store the url `/account/{id}` or something like that. This kind of filter should probably be applied from your site settings as some sort of `URL rollup rules` or whatever we want to call it.

**Reply by ukutaht** [reply]
We could make such a persistent filter apply backwards to historical data as well. But in that case you would lose the granularity of the URLs and you could never restore it.

**Reply by metmarkosaric** [reply]
yeah this is what i meant. i just want to search the data to give me more filtering options to analyse the traffic, i don't want to change anything in the data that is stored. cloudflare/ga/search console pretty much all have this same thing. click "manual filter", select if you want to include/exclude and then choose the metric and type what you want. i think the use case where you want to change the data stored is very rare compared to the one where you just want to analyse the data.

**Reply by captainrobbo** [reply]
Another approach occurred to me.   Would it make more sense to use (abuse?) custom goals and properties for this?   e.g. on pages relating to my competition, add

plausible('viewed_competition', {props: {competition: '/x/2021/EST/tallinn_marathon'}, language: 'ET'})

Sorry but I haven't really explored custom properties yet, so I don't know what kind of reports they produce.   But it seems to me the website developer will know how to classify a page when they serve it, and this might be simpler than having 'watch lists' of custom URL patterns.

**Reply by ukutaht** [reply]
Yes @captainrobbo and actually @Vigasaurus is landing a PR on master next week that would help you to do this even without custom props: https://github.com/plausible/analytics/pull/750. It's not the ideal solution but using pageview goals for this means you don't need to redeploy the script or anything like that.

**Reply by metmarkosaric** [reply]
@jaimeiniesta your specific request about tracking blog posts is now possible using the new pageview goals. see how we do it on [our own live demo](https://plausible.io/plausible.io/pages?period=30d&goal=Visit+%2Fblog**) and read [how to do it yourself](https://plausible.io/docs/pageview-goals).

**Reply by jaimeiniesta** [reply]
Thanks! Just tried it and it's exactly what I was needing 💪

**Comment by captainrobbo** [comment]
I see the PR - can't comment on implementation as I am new to Elixir, but
this looks absolutely perfect for our needs!    We are happy to test
aggressively as soon as it is there, and make our dashboard public.

On Fri, 26 Feb 2021 at 09:01, Uku Taht <notifications@github.com> wrote:

> Yes @captainrobbo <https://github.com/captainrobbo> and actually
> @Vigasaurus <https://github.com/Vigasaurus> is landing a PR on master
> next week that would help you to do this even without custom props: #750
> <https://github.com/plausible/analytics/pull/750>. It's not the ideal
> solution but using pageview goals for this means you don't need to redeploy
> the script or anything like that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/plausible/analytics/discussions/136#discussioncomment-408424>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACVQ5ZSN56SS624I5YZQ6D3TA5PNPANCNFSM4YELLYIQ>
> .
>


-- 
Andy Robinson
Managing Director, ReportLab
Wimbletech Zone 2
35 Wimbledon Hill Road
London SW19 7NB, UK
Tel +44-20-8191-7277

**Comment by captainrobbo** [comment]
A followup:  I found it quite easy to use custom goals to pull out total hits on each competition we're hosting.  Each of these subtotals represents hits on a whole forest of different pages in the last 30min, possibly on different subdomains and with different language prefixes.  I'm sure the ability to process URLs as you ingest them will be great, but this is an easy way to achieve the same thing.

![custom_goals](https://user-images.githubusercontent.com/11210470/109357290-e39c9d80-7879-11eb-80de-a0284b397582.png)

Plausible is awesome!

**Comment by ffont** [comment]
Any news on this? We're interested in that feature too. We can use Goals as a temporary workaround I guess, but without aggregation of pages the "Top pages" statistic is useless and consuming unnecessary resources.
Thanks a lot, so far Plausible has been great!

**Comment by scratchyone** [comment]
This would be very helpful for me too, and currently goals don't work correctly for this because I still end up seeing user's various uuids in the top pages list, which ends up de-anonymizing the data and gives me access to people's pages that are unlisted.

**Comment by tiblu** [comment]
When it comes to the initial use case of aggregating data, I FEEL like sending RAW data to Plausible and then Plausible UI enabling fancier aggregation functionality is the way to go. The RAW-er the data, the more into detail you can go processing it. IF you abstract the data, you lose level of detail. Never know what you want to query next. BUT, downside is that it makes Plausible UI more complicated and it is quite a development.

When it comes to privacy AND aggregation, I FEEL like ability to filter/modify the data that is being sent to Plausible is good for all parties. Implementation? Some kind of interceptor interface so User implemented function(method) gets called with data to be sent for modification.

**Reply by ukutaht** [reply]
Yeah I think we're in agreement. An ad-hoc query interface on top of the raw data would be most useful to all and we should build that first.

On-disk aggregation is necessary if you have personal data in URLs that you want to completely get rid of from your analytics. 

> Some kind of interceptor interface so User implemented function(method) gets called with data to be sent for modification.

Uh yeah I hadn't thought of this option as well. I assume you mean some sort of middleware on client-side? That would be super flexible.

**Reply by tiblu** [reply]
Yes, some kind of middleware on the client-side sounds good to me. I think its a low hanging fruit, relatively easy to implement and would benefit many. I have no specific ideas on the implementation tho.

**Reply by ukutaht** [reply]
> would benefit many

Potentially. Most of our users are marketers not programmers. An implementation that doesn't require coding from the user would be much more popular, albeit less flexible. Luckily most transformations that people want to make can be expressed with simple page rules like `/user/account/*`. These kinds of filters/aggregations could be built into the backend so you don't have to code on the frontend.

**Reply by scratchyone** [reply]
Maybe both of these? Since being able to remove personal data on the client side is quite important but it’d also be very useful to be able to aggregate data with the UI for anything that isn’t PII

**Reply by ukutaht** [reply]
Yeah, I'm in favour of offering all of these while prioritizing solutions that are most accessible to the widest section of our users possible.

**Comment by mschoeffmann** [comment]
I agree with the exact same situation - especially for PWA sites.

Two cases:
1. as mentioned above:
`/account/550e8400-e29b-11d4-a716-446655440000` >>> `/account/{uuid}`
2. aggregate sub-pages/actions as well
`/account/11111111-1111-1111-1111-111111111111/action1` >>> `/account/{uuid}/action1`
`/account/22222222-2222-2222-2222-222222222222/action1` >>> `/account/{uuid}/action1`
`/account/22222222-2222-2222-2222-222222222222/action2` >>> `/account/{uuid}/action2`

This would be really helpful, because sometimes you want to count how often action1 happened, not action1 per uuid ...

At least masking the ID's before counting would help?

**Reply by appinteractive** [reply]
Any update on that?

**Reply by metmarkosaric** [reply]
have you tried our custom filter? it doesn't overwrite the URLs but it does allow you to group your pages using wildcards. see https://plausible.io/docs/filters-segments

**Reply by appinteractive** [reply]
> have you tried our custom filter? it doesn't overwrite the URLs but it does allow you to group your pages using wildcards. see https://plausible.io/docs/filters-segments

Hey @metmarkosaric yeah I saw it on the screenshot but its not in the latest build (1.3.0 from April). So my question relates to the self hosted version I should say.

**Comment by thinkplaymake** [comment]
+1 on all of this, either post-capture aggregation, or the ability to track by a path you set in the path, i.e

/invite/hash1901892038910238920/view
/invite/hash2189382190389021921/view
/invite/hash1283902193021890382/view

I'd only want to track this as "/invite/view" (as an example), rather than hundreds of unique pageviews which are all effectively a single page view.

so i'd call something like:
plausible.track('/invite/view');

this might also solve for all the people asking for querystring support too, as they could call a custom page track.

**Comment by EmilePerron** [comment]
I would love to use this feature as well!

This could be implemented via some sort of client-side middleware as others have pointed out. 

Actually, from my point of view, I feel like it could be done fairly simply. 

I believe allowing for a location property/getter on `window.plausible` (and falling back on `location.href` when absent) would be enough to let developers add the implementation they need.

For example, if one wanted to simply specify a specific string as their location at load time, they could do something like this:
```js
window.plausible.location = "/my/redacted/route";
```

If they needed a more complex implementation, they could define `location` as a getter instead:
```js
Object.defineProperty(window.plausible, "location", {
  get: function() { 
    var url = window.location.href;

    // Custom logic that may or may not update `url` here

    return url;
  }
});
```


I'm not familiar with the Plausible codebase, but from a quick search it looks like this could be implemented by changing a single line in [plausible.js at line 51](https://github.com/plausible/analytics/blob/master/tracker/src/plausible.js#L51): 
```js
payload.u = window.plausible.location || location.href
```

Obviously, this would also require that the feature be documented and that test cases be written, but it would still be a minor addition that is backwards compatible and quick to implement for both the Plausible team and the developers who would end up using it.

What do you all think?

I'd be happy to make a PR for this if the concept is approved.

**Reply by ukutaht** [reply]
@EmilePerron  Thanks for your proposal!

I like your approach and it feels natural to me as well. Definitely a very flexible approach and much simpler than doing something on server-side.

Some thoughts that come to mind: We are totally obsessive about the size of the script. Anything that increases the size and is not used by default for everyone should be implemented as a [script extension](https://plausible.io/docs/script-extensions) instead.

The only added weight in your proposal would be this line:
```js
payload.u = window.plausible.location || location.href
```

However I don't think this would work if `window.plausible.location` is defined as a function, right? So it would have to check the type and decide whether to call the function, use the plain value or fall back to `location.href`.

The other thing I noticed is that your proposal expects  `window.plausible` to be defined. For that  you need the second line as shown here: https://plausible.io/docs/custom-event-goals. Not a huge deal but it feels sub-optimal if the user doesn't care about custom events, for example.

Thinking about these points I came up with a slightly modified API inspired by your proposal. If you want to provide a plain value (for ex. using the templating language from a server-generated page) you can use the standard configuration syntax:

```html
<script data-domain="my-domain.com" data-location="https://my-domain.com/my-custom-location" src="https://plausible.io/js/plausible.custom-location.js"></script>
```

And if you want to generate it client-side:
```html
<script>window.myLocationGetter = function() { return location.href + 'foo';  }</script>
<script data-domain="my-domain.com" data-get-location="myLocationeGetter" src="https://plausible.io/js/plausible.custom-location.js"></script>
```

The contract would be that `data-get-location` will provide the name of the function and if it exists of `window`, it will get executed and the return value used as the `location` value.

What do you think?

**important** We have to think about how this is going to play together with this request: https://github.com/plausible/analytics/discussions/177. It's not obvious at all how to satisfy both with a clean API. I think we have to address both requests in one go in order not to paint ourselves into a corner. I didn't try to reconcile the two in this comment.

**Reply by EmilePerron** [reply]
@ukutaht Thanks for the speedy response!

My bad on the on the possibly missing `window.plausible`. I do like the data attribute configuration better: it's clean and simple. 

And I get your obsession on file size: it's one of the things that makes Plausible Plausible, so better not break it!
Seing as this is an extra feature that many won't use, it makes sense to exclude it from the default script. So script extension it is.

**Now, about compatibility with #177:**
From what I can see, the URL that is sent by the tracker is the full URL, query string included. 
The query string trimming is done server-side in `PlausibleWeb.Api.ExternalController::create_event()`.

I think it's safe to assume that the implementation of the query parameter whitelist won't change this basic behaviour. It would likely be a script extension that adds a flag (or an array of whitelisted query parameters) to the payload, and the rest would still be handled server-side. Or, the whitelist would be handled in Plausible's dashboard, removing the need for a script extension. Either way, the basic behaviour described above stays the same: full URL is sent, and the server does the query filtering.

With that in mind, the only issue I can think of is that if the URL you specify via `data-location`/`data-get-location` contains a query string, it will always be trimmed just like URLs provided by the default behaviour. So there could be a discrepancy between your specified input and the data in Plausible.

However, seeing as query string trimming is already the default, this is not a new issue: it's still just #177. So when the query parameter whitelist (or other alternative) is eventually implemented, it will also work for the custom URLs introduced in this issue.

The only alternative I could think of would be to consider all query parameters in custom URLs as whitelisted, but this is likely to cause more issues than it solves, as it would shift the query-trimming responsibility to the users.

Let me know if I'm missing anything or if you have any more thoughts on the subject.

Cheers! 🤓

**Reply by EmilePerron** [reply]
@ukutaht Just pinging you to see if you have any more thoughts / oppositions to this. If not, I'll  get to work on a PR in the upcoming days!

**Reply by ukutaht** [reply]
Hey,

I've thought about this a bit and I think your proposal sounds good. Compatibility with #177 isn't an issue because I think the query param allowlist is better implemented on the backend and integrated into our pipeline there.

**Reply by EmilePerron** [reply]
I've started to work on this in #1398 - interested parties can keep an eye on the development / progress there :)

**Comment by metmarkosaric** [comment]
This is now live! 🔥

You can now filter page URLs and specify custom locations. This is especially helpful to redact and aggregate multiple pages whose URLs contain identifiers specific to users. Learn how it works here: [Specify a custom location to aggregate pages that contain identifiers](https://plausible.io/docs/custom-locations).

Thanks to @EmilePerron for the contribution!

**Comment by ffont** [comment]
Hi everyone,

I've been trying to use this feature but I don't seem to be able to get it working. Should it be working with a self-hosted 1.4.4 plausible? I'm following what's written in this tutorial (https://plausible.io/docs/custom-locations) but still see the pageviews without the custom location.

Thanks!

**Reply by EmilePerron** [reply]
Hey @ffont!

Just to be sure, are you looking at your existing data or at new data coming in after having set up custom locations?  
Existing data will not be changed/updated when you implement a custom location; it only affects new pageviews and events that are recorded after the change.

If the new data coming in is still not working with the custom location, I'd be happy to help if you can provide a snippet of what your code looks like for the Plausible scripts (including the custom location).

Cheers! :)

**Reply by ffont** [reply]
Hi @EmilePerron, thanks for your answer!
I'm looking at new data indeed, but it does not seem to be working. Here is my code:

```
<script defer data-domain="freesound.org" src="https://analytics.freesound.org/js/plausible.manual.js"></script>
<script>window.plausible = window.plausible || function() { (window.plausible.q = window.plausible.q || []).push(arguments) }</script>
<script>
  let url = window.location.href;
  let redactedUrl = url.replace(/\/\d+\//g, "/_ID_/"); // Replace numberic IDs
  redactedUrl = redactedUrl.replace(/people\/(\w+)\//ig, "people/_ID_/");  // Replace usernames
  redactedUrl = redactedUrl.replace(/browse\/geotags\/(\w+)\//ig, "browse/geotags/_TAG_/");  // Replace tag frm geotags
  redactedUrl = redactedUrl.replace(/browse\/tags\/.*/ig, "browse/tags/_TAGS_/");  // Replace multiple tags form tags
  redactedUrl = redactedUrl.replace('freesound.org/', 'freesound.org/NG/'); // Alter path to indicate NG front end
  plausible('pageview', { u: redactedUrl });
</script>
```

To me it acts like if I was not passing the `u` parameter to `plausible(...)` because I do see pageviews being shown but not with the rewritten URLs.

**Reply by EmilePerron** [reply]
I just checked out your version of `plausible.manual.js` (at https://analytics.freesound.org/js/plausible.manual.js), and it's different than the official one (https://plausible.io/js/plausible.manual.js). 

Amongst some of the differences, yours does not include the `u` check, as you mentioned. 

The official one checks as so: `n.u = e && e.u ? e.u : r.href`  
Your version checks like this: `n.u = r.href`

Is it possible that you're using an old Docker image of `plausible/analytics`?

**Reply by ffont** [reply]
Hi @EmilePerron,

Theoretically we use the official docker image `plausible/analytics:1.4.4`. Is this the correct one or we messed up somehow?

**Reply by EmilePerron** [reply]
It looks like the 1.4.4 tag (which was created a month ago) does not include the custom locations, which were merged and launched ~ 3 months ago. Weird.

@ukutaht Are some features not included in the docker images for Plausible? 

(perhaps it would be worth opening a separate issue/discussion for this).

**Reply by ukutaht** [reply]
> @ukutaht Are some features not included in the docker images for Plausible?

Self-hosted is released twice a year or so. Please see the [changelog](https://github.com/plausible/analytics/blob/master/CHANGELOG.md) for what is released and what is going in 1.5.

**Reply by ffont** [reply]
Thanks for the clarification @ukutaht ,
So considering 1.4.0 was released in October 2021, should we expect 1.5 around April?
Is it also a possibility to build our own image with the Dockerfile from master branch if we want to get the feature before the official release? Would it be advised?
Thanks again!

**Reply by ukutaht** [reply]
>  So considering 1.4.0 was released in October 2021, should we expect 1.5 around April?

Yes

> Is it also a possibility to build our own image with the Dockerfile from master branch if we want to get the feature before the official release? Would it be advised?

It's possible but you're on your own :) we have made some config changes which aren't documented for self-hosters so likely it would break. Better to wait until we can cut a release with docs on how to install, configure and upgrade it.

**Reply by ffont** [reply]
Thanks @ukutaht, we'll wait for the next release!
